src/torch_ucc.cpp

```diff
@@ -718,6 +737,88 @@ ProcessGroupUCC::~ProcessGroupUCC() {
   }
 }
 
+void ProcessGroupUCC::runHealthCheck() {
+  // Run health check in a separate thread and wait on CV to handle timeouts.
+  // This design allows us to handle hangs.
+
+  // When size_ is 1, there is no need to do any communication at all.
+  if (size_ == 1) return;
+
+  struct HealthCheckData {
+    std::mutex healthCheckMutex;
+    std::condition_variable healthCheckCv;
+    bool ucxHealthCheckSuccess = false;
+    bool uccHealthCheckSuccess = false;
+    std::exception_ptr healthCheckException;
+  } healthCheckData;
+  auto t = std::thread([&healthCheckData, this]() {
+    try {
+      auto oob = std::make_shared<torch_ucc_oob_coll_info_t>();
+      oob->rank = this->oob->rank;
+      oob->size = this->oob->size;
+      oob->store = this->oob->store;
+
+      std::vector<ucp_ep_h> eps;
+      ucc_team_h team = nullptr;
+      uint32_t comm_id;
+
+      // auto comm = std::make_unique<CommPG>(logger, oob, c10::kCPU, true);//comm_id, c10::kCPU, oob, logger, true);
+      auto comm = CommPG::get_comm(comm_id, c10::kCPU, oob, logger, true);
```

Comment:
Can we have an option to test comm on CPU or GPU or both?

I guess the challenging part is we might not know which GPU to use at this point? For GPU health check, perhaps we can just choose one GPU based on local_rank, and the communicator will be released/destroyed anyway once we finish the health check.

Reply:
using one GPU - that wouldn't provide full test coverage right? how else can we solve this?

Reply:
ProcessGroupNCCL uses the following to get device index for health check:
```C++
// Return CUDA device with ordinal given by input rank.
at::Device getDeviceForRank(int rank) {
  TORCH_CHECK(rank >= 0, "Invalid rank ", rank);
  auto numGPUs = at::cuda::getNumGPUs();
  int16_t deviceIdx = static_cast<int16_t>(rank % numGPUs);
  return at::Device(at::DeviceType::CUDA, deviceIdx);
}
```
I think we can match the same behavior.

Reply:
> using one GPU - that wouldn't provide full test coverage right? how else can we solve this?

@srinivas212  - I mean one GPU per process (choose based on local rank) like NCCL PG does as @zasdfgbnm mentioned. I think it covers most common use cases.

Reply:
Added GPU health check using the same device as NCCL does